scapy/fields.py

```diff
@@ -2155,3 +2155,77 @@ def i2repr(self, pkt, x):
+
+
+class ScalingField(Field):
+    """ Handle physical values which are scaled and/or offset for communication
+
+       Example:
+           >>> from scapy.packet import Packet
+           >>> class ScalingFieldTest(Packet):
+                   fields_desc = [ScalingField('data', 0, scaling=0.1, offset=-1, unit='mV')]  # noqa: E501
+           >>> ScalingFieldTest(data=10).show2()
+           ###[ ScalingFieldTest ]###
+             data= 10.0 mV
+           >>> hexdump(ScalingFieldTest(data=10))
+           0000  6E                                               n
+           >>> hexdump(ScalingFieldTest(data=b"\x6D"))
+           0000  6D                                               m
+           >>> ScalingFieldTest(data=b"\x6D").show2()
+           ###[ ScalingFieldTest ]###
+             data= 9.9 mV
+
+        bytes(ScalingFieldTest(...)) will produce 0x6E in this example.
+        0x6E is 110 (decimal). This is calculated through the scaling factor
+        and the offset. "data" was set to 10, which means, we want to transfer
+        the physical value 10 mV. To calculate the value, which has to be
+        sent on the bus, the offset has to subtracted and the scaling has to be
+        applied by division through the scaling factor.
+        bytes = (data - offset) / scaling
+        bytes = ( 10  -  (-1) ) /    0.1
+        bytes =  110 = 0x6E
+
+        If you want to force a certain internal value, you can assign a byte-
+        string to the field (data=b"\x6D"). If a string of a bytes object is
+        given to the field, no internal value conversion will be applied
+
+       :param name: field's name
+       :param default: default value for the field
+       :param scaling: scaling factor for the internal value conversion
+       :param unit: string for the unit representation of the internal value
+       :param offset: value to offset the internal value during conversion
+       :param ndigits: number of fractional digits for the internal conversion
+       :param fmt: struct.pack format used to parse and serialize the internal value from and to machine representation # noqa: E501
+       """
+    __slots__ = ["scaling", "unit", "offset", "ndigits"]
+
+    def __init__(self, name, default, scaling=1, unit="",
+                 offset=0, ndigits=3, fmt="B"):
+        self.scaling = scaling
+        self.unit = unit
+        self.offset = offset
+        self.ndigits = ndigits
+        Field.__init__(self, name, default, fmt)
+
+    def i2m(self, pkt, x):
+        if x is None:
+            x = 0
+        x = (x - self.offset) / self.scaling
+        if isinstance(x, float):
+            x = int(round(x))
+        return x
+
+    def m2i(self, pkt, x):
+        x = x * self.scaling + self.offset
+        if isinstance(x, float):
+            x = round(x, self.ndigits)
+        return x
+
+    def any2i(self, pkt, x):
+        if isinstance(x, str) or isinstance(x, bytes):
+            x = struct.unpack(self.fmt, raw(x))[0]
```

Comment:
Originally, `raw()` was only made for Packets:
`raw(IP())`
but then as it was handy, it became also used (pretty much only in scapy) to handle strings.

This is confusing and pretty inefficient (the calls are much slower as based on try/expect). https://github.com/secdev/scapy/pull/1735 will break `raw()` handling strings. We'll have `bytes_encode` for that instead (clearer right ? 😄)

That's a bit misleading, you can either work around it for now or wait for https://github.com/secdev/scapy/pull/1735 to be merged and use `bytes_encode`

Reply:
Thanks for this information. I'll work around it for now.